mscore/metaedit.cpp

```diff
@@ -19,94 +19,294 @@
-         tr("New tag name:")
-         );
-      QGridLayout* grid = static_cast<QGridLayout*>(scrollWidget->layout());
-      if (!s.isEmpty()) {
-            int idx = grid->rowCount();
-            QLabel* label = new QLabel;
-            label->setText(s);
-            QLineEdit* text = new QLineEdit;
-            grid->addWidget(label, idx, 0);
-            grid->addWidget(text, idx, 1);
-            }
-      dirty = true;
+      QLineEdit* tag = new QLineEdit;
+      tag->setPlaceholderText(tr("Tag name"));
+      connect(tag, SIGNAL(textChanged(const QString&)), SLOT(setDirty()));
+
+      QLineEdit *value = new QLineEdit;
+      value->setPlaceholderText(tr("Tag value"));
+      connect(value, SIGNAL(textChanged(const QString&)), SLOT(setDirty()));
+
+      QToolButton *deleteButton = new QToolButton;
+      deleteButton->setIcon(*icons[int (Icons::close_ICON)]);
+      deleteButton->setIconSize(QSize(12, 12));
+
+      connect(deleteButton, SIGNAL(clicked()), SLOT(setDirty()));
+      connect(deleteButton, SIGNAL(clicked()), tag,          SLOT(deleteLater()));
+      connect(deleteButton, SIGNAL(clicked()), value,        SLOT(deleteLater()));
+      connect(deleteButton, SIGNAL(clicked()), deleteButton, SLOT(deleteLater()));
+
+      int numFlags = scrollAreaLayout->rowCount();
+      scrollAreaLayout->addWidget(tag,          numFlags, 0);
+      scrollAreaLayout->addWidget(value,        numFlags, 1);
+      scrollAreaLayout->addWidget(deleteButton, numFlags, 2);
+
+      setDirty(true);
+      tag->setFocus();
+      // scroll down to see the newly created tag.
+      // the following line doesn't work for some reason.
+      //scrollArea->verticalScrollBar()->triggerAction(QAbstractSlider::SliderToMaximum);
+      QScrollBar* scrbar = scrollArea->verticalScrollBar();
+      int max = 100000;
+      scrbar->setMaximum(max);
+      scrbar->setValue(max);
       }
 
 //---------------------------------------------------------
-//   accept
+//   setDirty
 //---------------------------------------------------------
 
-void MetaEditDialog::accept()
+void MetaEditDialog::setDirty(bool dirty)
       {
+      if (dirty == m_dirty)
+            return;
+
       if (dirty) {
-            QGridLayout* grid = static_cast<QGridLayout*>(scrollWidget->layout());
-            int idx = grid->rowCount();
-            QMap<QString, QString> m;
+            saveButton->setEnabled(true);
+            setWindowTitle("Score properties: " + m_score->title() + tr("*", "For unsaved changes. This is probably useless,"
+                                                                             "but hey, I have no idea if there's asterisks in"
+                                                                             "cyrillic and arabic."));
+            }
+      else {
+            saveButton->setEnabled(false);
+            setWindowTitle("Score properties: " + m_score->title());
+            }
+
+      m_dirty = dirty;
+      }
```

Comment:
Here's where m_dirty is set to dirty, in function setDirty.

Reply:
But that is not the same `dirty' as the one that you defined locally in `setDirty()`. You need to drop the `bool` there

Reply:
What? I didn't get your idea, Jojo. `setDirty` uses default value (`true`) for the `dirty` variable, so that calling `setDirty()` w/o arguments will lead to calling `setDirty(true)`. Then, the `m_dirty = dirty` assignment happens here.

Reply:
Outsh! I missed the change from `{}` to `()`, the change from an inline function to a declaration with a default value sorry!